src/ppx_deriving.mli

```diff
@@ -18,15 +18,34 @@ open Parsetree
     annotations. If this function is missing, the corresponding [[%foo:]]
     annotation is ignored. *)
 type deriver = {
+  name : string ;
   core_type : (core_type -> expression) option;
-  structure : options:(string * expression) list -> path:string list ->
-              type_declaration list -> structure;
-  signature : options:(string * expression) list -> path:string list ->
-              type_declaration list -> signature;
+  type_decl_str : options:(string * expression) list -> path:string list ->
+                   type_declaration list -> structure;
+  type_ext_str : options:(string * expression) list -> path:string list ->
+                   type_extension -> structure;
+  type_decl_sig : options:(string * expression) list -> path:string list ->
+                   type_declaration list -> signature;
+  type_ext_sig : options:(string * expression) list -> path:string list ->
+                  type_extension -> signature;
 }
 
-(** [register name deriver] registers [deriver] as [name]. *)
-val register : string -> deriver -> unit
+(** [register deriver] registers [deriver] according to its [name] field. *)
+val register : deriver -> unit
+
+(** Creating {!deriver} structure. *)
+val create :
+  string ->
+  ?core_type: (core_type -> expression) ->
+  ?type_ext_str: (options:(string * expression) list -> path:string list ->
+                   type_extension -> structure) ->
+  ?type_ext_sig: (options:(string * expression) list -> path:string list ->
+                   type_extension -> signature) ->
+  type_decl_str: (options:(string * expression) list -> path:string list ->
```

Comment:
I can imagine derivers that only want to derive core types, or derivers that never work in signatures.

My point is, there's no conceptual difference between showing an error in `type_ext_str` and `type_decl_str`, why make one of these awkward?